lib/models/unfurl.js

```diff
@@ -0,0 +1,226 @@
+      const key = `${slackWorkspace.slackId}-${this.channelSlackId}-${this.url}`;
+      return cache.set(key, true);
+    },
+
+    async deliver() {
+      const slackUser = await SlackUser.findById(this.slackUserId);
+      const github = await this.constructor.getGitHubClient(this.channelSlackId, slackUser.slackId);
+      const {
+        attachment,
+        githubType,
+      } = await this.constructor.getAttachment(github, this.url, this.isCondensed, this.isPublic);
+      const unfurls = {
+        [this.url]: attachment,
+      };
+      logger.debug(unfurls, 'Unfurling links');
+      const slackWorkspace = await SlackWorkspace.findById(this.slackWorkspaceId);
+      const slackRes = await slackWorkspace.client.chat.unfurl({
+        ts: this.slackMessageTimestamp,
+        channel: this.channelSlackId,
+        unfurls,
+      });
+      logger.trace(slackRes, 'Unfurl complete');
+      await this.removeUnfurlEligibility();
+      this.githubType = githubType;
+      this.isDelivered = true;
+      await this.save();
+    },
+  });
+
+  Object.assign(Unfurl, {
+    async getGitHubClient(channelId, userSlackId) {
+      const github = new GitHubApi({ logger });
+
+      if (!process.env.ALLOWED_CHANNELS || !process.env.ALLOWED_CHANNELS.split(',').includes(channelId)) {
+        // keep the old way of doing things working
+        if (process.env.GITHUB_TOKEN) {
+          github.authenticate({
+            type: 'token',
+            token: process.env.GITHUB_TOKEN,
+          });
+          logger.debug({
+            channelId,
+          }, 'using PAT authed github client for unfurl');
+        }
+        return github;
+      }
+
+      const slackUser = await SlackUser.findOne({
+        where: { slackId: userSlackId },
+        include: [GitHubUser],
+      });
+
+      if (!slackUser.GitHubUser) {
+        throw Error('no github account connected');
+      }
+
+      github.authenticate({
+        type: 'oauth',
+        token: slackUser.GitHubUser.accessToken,
+      });
+      logger.debug({
+        channelId,
+        userSlackId: slackUser.slackId,
+        userGitHubId: slackUser.GitHubUser.id,
+      }, 'using user authed github client for unfurl');
+      return github;
+    },
+
+    async isEligibleForUnfurl(teamId, channelId, url) {
+      const key = `${teamId}-${channelId}-${url}`;
+      const recentlySeen = await cache.get(key);
+      return !recentlySeen;
+    },
+
+    async getAttachment(github, url, isCondensed, isPublic) {
+      const unfurlType = isCondensed ? 'condensed' : 'full';
+      const params = githubUrl(url);
+      if (!params || !resources[params.type]) {
+        throw new UnsupportedResource(url);
+      }
+      const attachment = await resources[params.type](params, github, unfurlType);
+
+      // Add lock to the footer if resource is not public
+      if (!isPublic) {
+        attachment.footer = `:lock: ${attachment.footer}`;
+      }
```

Comment:
Here is what this looks like. Feels kinda weird to do this outside of the renderer/messages folder but implementation wise this solution has far fewer lines of code.

![image](https://user-images.githubusercontent.com/7718702/38211546-c1d4f0dc-36b2-11e8-96a0-2853cef81d9e.png)